refactor!: move date-picker scrollers to light DOM

web-padawan opened the pull request:
## Description

Fixes #4287

## Changes 

### Overlay content

1. Added new `vaadin-date-picker-month-scroller` and `vaadin-date-picker-year-scroller` elements,
2. Changed both infinite scrollers to be placed in corresponding `vaadin-date-picker-overlay-content` slots,
3. Year scroller now uses `vaadin-date-picker-year` element with `year-number` and `year-separator` parts,
4. Changed `desktop` attribute to be set on the host, like `fullscreen` (instead of elements in shadow DOM),
5. Added `flush()` to fix  logic focusing date element on opening the overlay when `fullscreen` mode is used.

### Infinite scroller

1. Converted `InfiniteScroller` to a base class extended by new scrollers (without custom element definition),
2. Replaced Polymer `templatize` with `_createElement` + `_updateElement` functions and property observers,
3. Removed no longer needed `setTimeout` calls in `InfiniteScroller` class to reduce asynchronous logic,
4. Updated tests to not set custom `bufferSize` to scrollers (it doesn't seem to really affect performance).

## Before

![scrollers-before](https://user-images.githubusercontent.com/10589913/196669601-3dc99e0d-fea7-415e-b826-4d842f1b97fa.png)

## After

![scrollers-after](https://user-images.githubusercontent.com/10589913/196670282-1e9264fc-3bc8-4d1e-9e98-be3d445dfde2.png)

## Type of change

- Breaking change

web-padawan commented on packages/date-picker/src/vaadin-infinite-scroller.js:
```diff
@@ -315,43 +331,43 @@ class InfiniteScroller extends PolymerElement {
         itemWrapper.setAttribute('slot', slotName);
         this.appendChild(itemWrapper);
 
-        setTimeout(() => {
-          // Only stamp the visible instances first
-          if (this._isVisible(itemWrapper, container)) {
-            this._ensureStampedInstance(itemWrapper);
-          }
-        }, 1); // Wait for first reset
```
Using `setTimeout` here was added in https://github.com/vaadin/vaadin-date-picker/pull/268 to defer rendering of the scrollers until the overlay is opened. See https://github.com/vaadin/vaadin-date-picker/pull/268/commits/6709076f95ced32098f61357c6703bd7f517a623

Originally, the "lazy loading" was relying on animation events. However, later this logic was changed in https://github.com/vaadin/vaadin-date-picker/pull/297, see https://github.com/vaadin/vaadin-date-picker/commit/fff65dff7fb89a01590198c3674ac5d8e6b63e4b

After some debugging and testing, I don't see any reason for keeping `setTimeout` here. 
There is already `afterNextRender` used to render instances that aren't initially visible.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-mixin.js:
```diff
@@ -84,6 +84,7 @@ export const DatePickerMixin = (subclass) =>
          */
         showWeekNumbers: {
           type: Boolean,
+          value: false,
```
Added default value to avoid passing `undefined` to overlay content and month calendar elements.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-overlay-content.js:
```diff
@@ -64,62 +67,17 @@ class DatePickerOverlayContent extends ControllerMixin(ThemableMixin(DirMixin(Po
-        }
-
-        #scrollers[desktop] [part='months'] {
+        :host([desktop]) ::slotted([slot='months']) {
```
Changed to set `desktop` state attribute on the host, so that it can be used for styling.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-overlay-content.js:
```diff
@@ -309,15 +243,11 @@ class DatePickerOverlayContent extends ControllerMixin(ThemableMixin(DirMixin(Po
 
   get focusableDateElement() {
-    return this.calendars.map((calendar) => calendar.focusableDateElement).find(Boolean);
+    return this.calendars.map((calendar) => calendar.isConnected && calendar.focusableDateElement).find(Boolean);
```
Added `isConnected` because some tests that open overlay might exit yearly, without waiting for date element to be focused. So, after the promise that waits for them to render is resolved, this getter is called, and the error is thrown on accessing `vaadin-month-calendar` shadow DOM, because the calendar is already detached.

I will try to fix those tests instead of using this workaround (although we could keep it as well).

web-padawan replied:
Fixed some tests and removed `isConnected` check, it should not be needed anymore.

web-padawan commented on packages/date-picker/src/vaadin-infinite-scroller.js:
```diff
@@ -69,10 +68,6 @@ class InfiniteScroller extends PolymerElement {
     `;
   }
 
-  static get is() {
-    return 'vaadin-infinite-scroller';
-  }
```
Removed custom element definition because `InfiniteScroller` is kind of an abstract class now.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-overlay-content.js:
```diff
@@ -289,11 +209,26 @@ class DatePickerOverlayContent extends ControllerMixin(ThemableMixin(DirMixin(Po
   static get observers() {
-    return ['__updateCancelButton(_cancelButton, i18n)', '__updateTodayButton(_todayButton, i18n, minDate, maxDate)'];
+    return [
+      '__updateCalendars(calendars, i18n, minDate, maxDate, selectedDate, focusedDate, showWeekNumbers, _ignoreTaps, _theme)',
```
New observers are needed to propagate properties previously handled by templatizer `forwardHostProp`.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-overlay-content.js:
```diff
@@ -309,15 +244,11 @@ class DatePickerOverlayContent extends ControllerMixin(ThemableMixin(DirMixin(Po
-  }
-
-  get calendars() {
-    return [...this.shadowRoot.querySelectorAll('vaadin-month-calendar')];
```
Changed logic to only get calendars instances once, because in practice they are never re-rendered.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-overlay-content.js:
```diff
@@ -936,7 +969,11 @@ class DatePickerOverlayContent extends ControllerMixin(ThemableMixin(DirMixin(Po
     if (!this.calendars.length) {
       await new Promise((resolve) => {
-        setTimeout(resolve);
+        afterNextRender(this, () => {
+          // Force dom-repeat elements to render
+          flush();
+          resolve();
+        });
```
Replaced `setTimeout` with a more reliable logic to ensure date cells are actually rendered.
This is a proper fix for the issue that I previously attempted to fix in #4296.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-overlay-content.js:
```diff
@@ -502,17 +535,23 @@ class DatePickerOverlayContent extends ControllerMixin(ThemableMixin(DirMixin(Po
   }
 
   _initialPositionChanged(initialPosition) {
+    if (this._monthScroller && this._yearScroller) {
+      this._monthScroller.active = true;
+      this._yearScroller.active = true;
+    }
```
This was previously set in template using `active="[[initialPosition]]"` to trigger rendering of the scrollers.
We could consider refactoring this to use `afterNextRender()` instead of using the observer.

web-padawan commented on packages/date-picker/src/vaadin-infinite-scroller.js:
```diff
@@ -155,28 +144,52 @@ class InfiniteScroller extends PolymerElement {
       }
 
       this._initDone = true;
+      this.dispatchEvent(new CustomEvent('init-done'));
```
The event is used to notify overlay content so that we can get reference to calendars and years.
Could be probably refactored to use `afterNextRender` in the overlay content instead.

web-padawan commented on packages/date-picker/theme/material/vaadin-date-picker-overlay-content-styles.js:
```diff
@@ -88,51 +89,6 @@ registerStyles(
-    :host([years-visible]) [part='years']::after {
-      top: calc(20px + 16px);
-      height: calc(100% - 20px - 16px);
-    }
```
Removed in #4771.

web-padawan commented on packages/date-picker/src/vaadin-infinite-scroller.js:
```diff
@@ -116,26 +117,14 @@ class InfiniteScroller extends PolymerElement {
   ready() {
     super.ready();
 
-    this._buffers = Array.prototype.slice.call(this.root.querySelectorAll('.buffer'));
```
Fixed in #4777

web-padawan commented on packages/date-picker/src/vaadin-infinite-scroller.js:
```diff
@@ -155,28 +144,48 @@ class InfiniteScroller extends PolymerElement {
     if (!this._initDone) {
       // Once the first set of items start fading in, stamp the rest
       this._buffers.forEach((buffer) => {
-        [].forEach.call(buffer.children, (insertionPoint) => this._ensureStampedInstance(insertionPoint._itemWrapper));
```
Fixed in #4777

web-padawan commented on packages/date-picker/src/vaadin-date-picker-year-scroller.js:
```diff
@@ -0,0 +1,95 @@
+ * @mixes ThemableMixin
+ * @private
+ */
+class DatePickerYearScroller extends ThemableMixin(InfiniteScroller) {
```
I'm not sure whether we should make scrollers use `ThemableMixin` - in theory, corresponding styles could be still located in `vaadin-date-picker-overlay-content` instead. @jouni WDYT?

sissbruecker reviewed (commented):
Did an initial code review and basic testing. PR is in a really good shape IMO, only added a few minor comments.

After looking at the overlay content once more I was thinking that a lot of the date focus / scrolling logic could be encapsulated in the new scroller components. That would reduce the complexity of the overlay content component a bit and avoid doing scroller specific stuff (accessing calendars, scroll position calculations, detect sub-month scrolling) in the overlay. There would be some trade-offs as communication between overlay and scrollers would become more complex. Anyway, could be done at some point in the future, for now it's better to keep the changes minimal.

I still want to do some additional manual testing before approving.

web-padawan commented on packages/date-picker/theme/material/vaadin-date-picker-year-styles.js:
```diff
@@ -0,0 +1,28 @@
+      line-height: 10px; /* NOTE(platosha): chosen to align years to months */
+    }
+
+    [part='year-separator'] {
+      display: flex;
+      align-items: center;
+      justify-content: center;
+      height: calc(100% - 10px);
+    }
+
+    [part='year-separator']::before {
+      content: '';
+      background-color: currentColor;
+      width: 4px;
+      height: 4px;
+      border-radius: 50%;
+    }
```
Yes, I changed this exactly because of those visual glitches: with the new DOM structure, usage of `margin` was causing incorrect buffer height during the initial render, causing `position` of the year scroller to be off.

web-padawan commented on packages/date-picker/src/vaadin-date-picker-month-scroller.js:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  /** @private */
+  _dateAfterXMonths(months) {
```
Thanks, done. Also removed a test that was calling this private method 🙈

jouni commented:
Overall, I feel it’s quite fine to expose the new `<vaadin-date-picker-month-scroller>` and `<vaadin-date-picker-year-scroller>` elements. That just means those will be part of the public styling API from now no, and need to be maintained accordingly.

I suppose they are equally exposed as parts inside the `<vaadin-date-picker-overlay-content>` would be.

Should we use `ThemableMixin` for the new elements? I think no. We should not add new elements/components that use it, as it’s close to being deprecated. At least discouraged, and `::part() being encouraged.

web-padawan commented:
> Should we use ThemableMixin for the new elements? I think no.

Changed the new scrollers to not use `ThemableMixin` and moved corresponding CSS back to where it was.

web-padawan commented on packages/date-picker/theme/material/vaadin-date-picker-overlay-content-styles.js:
```diff
@@ -19,13 +20,10 @@ registerStyles(
       display: flex;
       align-items: baseline;
       position: relative;
-      z-index: 2;
```
Looks like everything works without `z-index` so I removed it.

web-padawan commented on packages/date-picker/theme/lumo/vaadin-date-picker-overlay-content-styles.js:
```diff
@@ -177,22 +143,19 @@ registerStyles(
+      ::slotted([slot='years']) {
         --vaadin-infinite-scroller-buffer-width: 90px;
         width: 50px;
+        background-color: var(--lumo-shade-5pct);
```
Moved this line here to avoid duplicate CSS selectors.

sissbruecker commented on packages/date-picker/theme/lumo/vaadin-date-picker-overlay-content-styles.js:
```diff
@@ -58,15 +56,8 @@ registerStyles(
-    [part='years']:hover [part='year-number'],
-    [part='years']:hover [part='year-separator'] {
-      opacity: 1;
+    ::slotted([slot='years']):hover {
+      --_lumo-date-picker-year-opacity: 1;
     }
```
This selector isn't working properly, this seems to work instead:
```
::slotted([slot='years']:hover)
```

web-padawan replied:
Good catch, fixed.
